File: src/main/java/ourvillagerdiscounts/ourvillagerdiscounts/mixin/VillagerInteractMixin.java
package ourvillagerdiscounts.ourvillagerdiscounts.mixin;

import net.minecraft.entity.passive.VillagerEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
import ourvillagerdiscounts.ourvillagerdiscounts.callback.VillagerInteractCallback;

@Mixin(VillagerEntity.class)
public class VillagerInteractMixin {
    @Inject(at = @At(value = "INVOKE"), method="interactMob", cancellable = true)
    private void onVillagerInteract(final PlayerEntity player, final Hand hand, final CallbackInfoReturnable<Boolean> info) {
        ActionResult result = VillagerInteractCallback.EVENT.invoker().interact(player, (VillagerEntity)(Object)this);

        if (result == ActionResult.FAIL) {
            info.cancel();
        }
    }
}

File: src/main/kotlin/ourvillagerdiscounts/ourvillagerdiscounts/mixin/VillagerInteractMixin.kt
package ourvillagerdiscounts.ourvillagerdiscounts.mixin

import org.spongepowered.asm.mixin.Mixin
import net.minecraft.entity.passive.VillagerEntity
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.At
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.Hand
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import net.minecraft.util.ActionResult
import ourvillagerdiscounts.ourvillagerdiscounts.callback.VillagerInteractCallback

@Mixin(VillagerEntity::class)
class VillagerInteractMixin {
    @Inject(at = [At(value = "INVOKE")], method = ["interactMob"], cancellable = true)
    private fun onVillagerInteract(player: PlayerEntity, hand: Hand, info: CallbackInfoReturnable<Boolean>) {
        val result = VillagerInteractCallback.EVENT.invoker().interact(player, this as VillagerEntity)
        if (result == ActionResult.FAIL) {
            info.cancel()
        }
    }
}